sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala

```diff
@@ -589,18 +590,40 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
         col.getType.startsWith(serdeConstants.CHAR_TYPE_NAME))
       .map(col => col.getName).toSet
 
-    filters.collect {
-      case op @ BinaryComparison(a: Attribute, Literal(v, _: IntegralType)) =>
-        s"${a.name} ${op.symbol} $v"
-      case op @ BinaryComparison(Literal(v, _: IntegralType), a: Attribute) =>
-        s"$v ${op.symbol} ${a.name}"
-      case op @ BinaryComparison(a: Attribute, Literal(v, _: StringType))
-          if !varcharKeys.contains(a.name) =>
-        s"""${a.name} ${op.symbol} ${quoteStringLiteral(v.toString)}"""
-      case op @ BinaryComparison(Literal(v, _: StringType), a: Attribute)
-          if !varcharKeys.contains(a.name) =>
-        s"""${quoteStringLiteral(v.toString)} ${op.symbol} ${a.name}"""
-    }.mkString(" and ")
+    def isExtractable(expr: Expression): Boolean =
+      expr match {
+        case Literal(_, _: IntegralType) | Literal(_, _: StringType) => true
+        case _ => false
+      }
+
+    def extractValue(expr: Expression): String =
+      expr match {
+        case Literal(v, _: IntegralType) => v.toString
+        case Literal(v, _: StringType) => quoteStringLiteral(v.toString)
+      }
+
+    lazy val convert: PartialFunction[Expression, String] =
+      {
+        case In(a: Attribute, exprs)
+            if !varcharKeys.contains(a.name) && exprs.forall(isExtractable) =>
+          val or =
+            exprs
+              .map(expr => s"${a.name} = ${extractValue(expr)}")
+              .reduce(_ + " or " + _)
+          "(" + or + ")"
+        case op @ BinaryComparison(a: Attribute, expr2)
+            if !varcharKeys.contains(a.name) && isExtractable(expr2) =>
+          s"${a.name} ${op.symbol} ${extractValue(expr2)}"
```

Comment:
Quoting the name causes the Hive metastore partition filter parser to throw an exception.

I don't think we need to add any more parens. Can you give a counterexample?